frame/contracts/src/exec.rs

```diff
@@ -307,31 +336,31 @@ where
 		input_data: Vec<u8>,
 	) -> ExecResult {
 		if self.depth == self.config.max_depth as usize {
-			Err("reached maximum depth, cannot make a call")?
+			Err(Error::<T>::MaxCallDepthReached)?
 		}
 
 		if gas_meter
 			.charge(self.config, ExecFeeToken::Call)
 			.is_out_of_gas()
 		{
-			Err("not enough gas to pay base call fee")?
+			Err(Error::<T>::OutOfGas)?
 		}
 
 		// Assumption: `collect_rent` doesn't collide with overlay because
 		// `collect_rent` will be done on first call and destination contract and balance
 		// cannot be changed before the first call
-		let contract_info = rent::collect_rent::<T>(&dest);
-
-		// Calls to dead contracts always fail.
-		if let Some(ContractInfo::Tombstone(_)) = contract_info {
-			Err("contract has been evicted")?
+		// We do not allow 'calling' plain accounts. For transfering value
+		// `ext_transfer` must be used.
+		let contract = if let Some(ContractInfo::Alive(info)) = rent::collect_rent::<T>(&dest) {
+			info
+		} else {
+			Err(Error::<T>::InvalidContractCalled)?
 		};
 
 		let transactor_kind = self.transactor_kind();
 		let caller = self.self_account.clone();
-		let dest_trie_id = contract_info.and_then(|i| i.as_alive().map(|i| i.trie_id.clone()));
 
-		self.with_nested_context(dest.clone(), dest_trie_id, |nested| {
+		self.with_nested_context(dest.clone(), Some(contract.trie_id.clone()), |nested| {
```

Comment:
Is there a case left where we pass `None` actually?

Reply:
You are right. I removed the Option.